File: Client/XyMusicForWin/src/application/use-cases/LibraryUseCases.ts
import type { LibraryRepository } from "../ports/LibraryRepository";
import type { FavoriteSort } from "../../domain/pagination";

export class LibraryUseCases {
  constructor(private readonly repository: LibraryRepository) {}

  favorites(sort: FavoriteSort, cursor?: string, limit = 50, signal?: AbortSignal) { return this.repository.getFavoriteTracks(sort, cursor, limit, signal); }
  history(cursor?: string, limit = 50, signal?: AbortSignal) { return this.repository.getHistoryTracks(cursor, limit, signal); }
  favorite(trackId: string, favorite: boolean, signal?: AbortSignal) {
    return this.repository.setFavorite(trackId, favorite, signal);
  }
}

File: Client/XyMusicForAndroid/domain/src/main/java/com/xymusic/app/feature/library/domain/LibraryUseCases.kt
package com.xymusic.app.feature.library.domain

import com.xymusic.app.feature.library.domain.model.FavoriteSort
import com.xymusic.app.feature.library.domain.model.PlaybackProgressCommand
import javax.inject.Inject

class LibraryUseCases
@Inject
constructor(private val repository: LibraryRepository) {
    fun observeIsFavorite(trackId: String) = repository.observeIsFavorite(trackId)

    fun favorites() = repository.favoriteTracks()

    fun history() = repository.playbackHistory()

    suspend fun refreshFavorites(sort: FavoriteSort = FavoriteSort.FAVORITED_DESC) = repository.refreshFavorites(sort)

    suspend fun refreshHistory() = repository.refreshHistory()

    suspend fun setFavorite(trackId: String, favorite: Boolean) = repository.setFavorite(trackId, favorite)

    suspend fun recordPlayback(command: PlaybackProgressCommand) = repository.recordPlayback(command)
}
